use-cases/conference-scheduling/src/main/java/org/acme/conferencescheduling/solver/ConferenceSchedulingConstraintProvider.java

```diff
@@ -0,0 +1,537 @@
+import static org.acme.conferencescheduling.domain.ConferenceConstraintConfiguration.TALK_MUTUALLY_EXCLUSIVE_TALKS_TAGS;
+import static org.acme.conferencescheduling.domain.ConferenceConstraintConfiguration.TALK_PREFERRED_ROOM_TAGS;
+import static org.acme.conferencescheduling.domain.ConferenceConstraintConfiguration.TALK_PREFERRED_TIMESLOT_TAGS;
+import static org.acme.conferencescheduling.domain.ConferenceConstraintConfiguration.TALK_PREREQUISITE_TALKS;
+import static org.acme.conferencescheduling.domain.ConferenceConstraintConfiguration.TALK_PROHIBITED_ROOM_TAGS;
+import static org.acme.conferencescheduling.domain.ConferenceConstraintConfiguration.TALK_PROHIBITED_TIMESLOT_TAGS;
+import static org.acme.conferencescheduling.domain.ConferenceConstraintConfiguration.TALK_REQUIRED_ROOM_TAGS;
+import static org.acme.conferencescheduling.domain.ConferenceConstraintConfiguration.TALK_REQUIRED_TIMESLOT_TAGS;
+import static org.acme.conferencescheduling.domain.ConferenceConstraintConfiguration.TALK_UNDESIRED_ROOM_TAGS;
+import static org.acme.conferencescheduling.domain.ConferenceConstraintConfiguration.TALK_UNDESIRED_TIMESLOT_TAGS;
+import static org.acme.conferencescheduling.domain.ConferenceConstraintConfiguration.THEME_TRACK_CONFLICT;
+import static org.acme.conferencescheduling.domain.ConferenceConstraintConfiguration.THEME_TRACK_ROOM_STABILITY;
+
+import java.time.LocalDate;
+import java.time.temporal.ChronoUnit;
+import java.util.Collections;
+import java.util.Objects;
+
+import ai.timefold.solver.core.api.score.stream.Constraint;
+import ai.timefold.solver.core.api.score.stream.ConstraintFactory;
+import ai.timefold.solver.core.api.score.stream.ConstraintProvider;
+
+import org.acme.conferencescheduling.domain.ConferenceConstraintConfiguration;
+import org.acme.conferencescheduling.domain.Speaker;
+import org.acme.conferencescheduling.domain.Talk;
+import org.acme.conferencescheduling.solver.justifications.SpeakerJustification;
+import org.acme.conferencescheduling.solver.justifications.TalkJustification;
+import org.acme.conferencescheduling.solver.justifications.TalkSpeakerJustification;
+import org.acme.conferencescheduling.solver.justifications.TwoTalkJustification;
+import org.acme.conferencescheduling.solver.justifications.TwoTalkSpeakerJustification;
+
+/**
+ * Provides the constraints for the conference scheduling problem.
+ * <p>
+ * Makes heavy use of CS expand() functionality to cache computation results,
+ * except in cases where doing so less is efficient than recomputing the result.
+ * That is the case in filtering joiners.
+ * In this case, it is better to reduce the size of the joins even at the expense of duplicating some calculations.
+ * In other words, time saved by caching those calculations is far outweighed by the time spent in unrestricted joins.
+ */
+public class ConferenceSchedulingConstraintProvider implements ConstraintProvider {
```

Comment:
Although I understand that doing a different justification for each of these constraints would have been crazy, maybe the current granularity is lower than it ought to be? 

At least `TalkTagJustification` would IMO be a good addition.